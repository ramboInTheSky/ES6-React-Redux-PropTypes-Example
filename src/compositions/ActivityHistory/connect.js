import { connect } from 'react-redux';
import path from 'ramda/src/path';

import { getActivityHistory, invalidateActivityHistory } from '../../ducks/activities';

export const mapStateToProps = state => {
  const {
    activityHistory: { heading, labels },
  } = state.dictionary;

  return {
    activityHistory: path(['activities', 'history'], state),
    heading,
    labels,
  };
};

export const mergeProps = (
  { arrears, tenants, vulnerabilityFlag, ...stateProps },
  { dispatch },
  ownProps
) => {
  const arrearsId = path(['match', 'params', 'arrearsId'], ownProps);

  return {
    ...stateProps,
    arrearsId,
    invalidateActivityHistory: () => dispatch(invalidateActivityHistory()),
    getActivityHistory: () => dispatch(getActivityHistory(arrearsId)),
    ...ownProps,
  };
};

export default connect(
  mapStateToProps,
  null,
  mergeProps
);
