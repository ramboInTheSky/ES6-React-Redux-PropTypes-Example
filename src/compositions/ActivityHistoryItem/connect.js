import { connect } from 'react-redux';
import path from 'ramda/src/path';

import activityNavigation from '../../util/activityNavigation';
import getHelpers from '../../util/stateHelpers';

import { updateRibbon } from '../../ducks/ribbon';
import { getActivity, getActivityHistory, invalidateActivityDetail } from '../../ducks/activities';

export const mapStateToProps = (state, ownProps) => {
  const { get, getBreadcrumbs } = getHelpers(state);

  const {
    activityHistory: { itemHeading: heading, labels, values, attachments: attachmentsHeading },
    arrearsDetails: { heading: arrearsDetailsHeading },
    genericErrorText,
    paymentPlan: { installmentArrangementFormat },
  } = state.dictionary;

  const referralLabels = path(['dictionary', 'referral'], state);
  const pauseLabels = path(['dictionary', 'pause'], state);
  const paymentLabels = path(['dictionary', 'payment', 'labels'], state);
  const paymentPlanLabels = path(['dictionary', 'paymentPlan', 'labels'], state);
  const legalReferralLabels = path(['dictionary', 'legalReferral'], state);

  const arrearsId = path(['match', 'params', 'arrearsId'], ownProps);
  const itemId = path(['match', 'params', 'itemId'], ownProps);
  const redirectToDetailsPage = () => ownProps.history.push(`/arrears-details/${arrearsId}`);
  const goToHistoryItem = historyId =>
    ownProps.history.push(`/arrears-details/${arrearsId}/history/${historyId}`);
  const activities = get(['activities', 'history'], []);
  const attachments = get(['activities', 'attachments']);
  const attachmentsLoading = get(['activities', 'attachmentsLoading'], []);
  const attachmentsError = get(['activities', 'attachmentsError'], []);
  const activityHistoryNavigation = activityNavigation(activities, itemId);

  return {
    activities,
    activityHistoryNavigation,
    genericErrorText,
    attachmentsHeading,
    activityDetail: get(['activities', 'detail']),
    activityError: get(['activities', 'error']),
    activityLoading: get(['activities', 'loading']),
    arrearsId,
    breadcrumb: getBreadcrumbs(
      heading,
      {},
      {
        label: arrearsDetailsHeading,
        to: `/arrears-details/${arrearsId}`,
      }
    ),
    canGoNext: !activityHistoryNavigation.isLast(),
    canGoPrevious: !activityHistoryNavigation.isFirst(),
    heading,
    goToHistoryItem,
    itemId,
    labels,
    redirectToDetailsPage,
    values,
    referralLabels,
    pauseLabels,
    paymentLabels,
    paymentPlanLabels: { installmentArrangementFormat, ...paymentPlanLabels },
    legalReferralLabels,
    attachments,
    attachmentsError,
    attachmentsLoading,
  };
};

export const mergeProps = (
  {
    activities,
    activityDetail,
    activityHistoryNavigation,
    arrearsId,
    breadcrumb,
    heading,
    goToHistoryItem,
    itemId,
    redirectToDetailsPage,
    ...stateProps
  },
  { dispatch },
  ownProps
) => ({
  ...stateProps,
  activities,
  itemId,
  activityDetail,
  getActivityById: id => {
    const activityHistoryItem = activities.find(activity => activity.id === id);
    dispatch(getActivity(activityHistoryItem)).catch(() => {});
  },
  getActivityHistory: () => {
    dispatch(getActivityHistory(arrearsId));
  },
  invalidateActivityDetail: () => dispatch(invalidateActivityDetail()),
  onBack: () => redirectToDetailsPage(),
  onNext: () => {
    const nextHistoryId = activityHistoryNavigation.getNextHistoryId();
    if (nextHistoryId) {
      goToHistoryItem(nextHistoryId);
    }
  },
  onPrevious: () => {
    const prevHistoryId = activityHistoryNavigation.getPreviousHistoryId();
    if (prevHistoryId) {
      goToHistoryItem(prevHistoryId);
    }
  },
  updatePageHeader: () => dispatch(updateRibbon({ title: heading, breadcrumb })),
  ...ownProps,
});

export default connect(
  mapStateToProps,
  null,
  mergeProps
);
