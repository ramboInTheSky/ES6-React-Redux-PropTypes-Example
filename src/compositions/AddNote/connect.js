import { connect } from 'react-redux';
import path from 'ramda/src/path';

import getHelpers from '../../util/stateHelpers';
import { addNotification } from '../../ducks/notifications';
import { updateRibbon } from '../../ducks/ribbon';
import { addNoteToCase, clearNoteError } from '../../ducks/notes';

export const mapStateToProps = (state, ownProps) => {
  const { get, getBreadcrumbs } = getHelpers(state);
  const {
    addNote: { addNote, formError, notification, ...text },
    arrearsDetails: { heading: arrearsDetailsHeading },
  } = state.dictionary;

  const arrearsId = path(['match', 'params', 'arrearsId'], ownProps);
  const returnUrl = `/arrears-details/${arrearsId}`;
  const redirectToDetailsPage = () => ownProps.history.push(returnUrl);

  const redirectToAttachmentsPage = id =>
    ownProps.history.push(`${returnUrl}/note/${id}/attachments?redirectTo=${returnUrl}`);

  return {
    addNote,
    arrearsId,
    breadcrumb: getBreadcrumbs(
      addNote,
      {},
      {
        label: arrearsDetailsHeading,
        to: `/arrears-details/${arrearsId}`,
      }
    ),
    error: get(['notes', 'error']),
    isLoading: get(['notes', 'loading']),
    formError,
    heading: addNote,
    notification,
    redirectToDetailsPage,
    redirectToAttachmentsPage,
    ...text,
  };
};

export const mergeProps = (
  {
    arrearsId,
    breadcrumb,
    heading,
    notification,
    redirectToAttachmentsPage,
    redirectToDetailsPage,
    ...stateProps
  },
  { dispatch },
  ownProps
) => ({
  ...stateProps,
  arrearsId,
  clearError: () => dispatch(clearNoteError()),
  onBack: redirectToDetailsPage,
  onSubmit: (note, wantsToAttachFiles) =>
    dispatch(addNoteToCase(arrearsId, note))
      .then(id => {
        dispatch(
          addNotification({
            dataBddPrefix: 'addNote',
            lines: [{ text: notification.line1 }],
            icon: 'success',
            notificationType: 'confirmation',
            title: notification.title,
          })
        );

        if (wantsToAttachFiles) {
          redirectToAttachmentsPage(id);
        } else {
          redirectToDetailsPage();
        }
      })
      .catch(() => {}),
  updatePageHeader: () => dispatch(updateRibbon({ title: heading, breadcrumb })),
  ...ownProps,
});

export default connect(
  mapStateToProps,
  null,
  mergeProps
);
