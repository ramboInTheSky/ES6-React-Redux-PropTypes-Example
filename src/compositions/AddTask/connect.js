import { connect } from 'react-redux';
import path from 'ramda/src/path';
import { formatting } from 'nhh-styles';
import getHelpers from '../../util/stateHelpers';
import { addNotification } from '../../ducks/notifications';
import { updateRibbon } from '../../ducks/ribbon';
import { clearTaskError, taskError, addTask as AddTaskItem } from '../../ducks/tasks';
import { clearSearch, search } from '../../ducks/housingofficersearch';

export const mapStateToProps = (state, ownProps) => {
  const { get, getBreadcrumbs } = getHelpers(state);
  const {
    addTask,
    addTask: { addTask: heading, notification, formError, serverError },
    arrearsDetails: { heading: arrearsDetailsHeading },
  } = state.dictionary;

  const arrearsId = path(['match', 'params', 'arrearsId'], ownProps);
  const redirectToDetailsPage = () => ownProps.history.push(`/arrears-details/${arrearsId}`);

  return {
    addTask,
    arrearsId,
    breadcrumb: getBreadcrumbs(
      heading,
      {},
      {
        label: arrearsDetailsHeading,
        to: `/arrears-details/${arrearsId}`,
      }
    ),
    loading: get(['tasks', 'loading']),
    housingOfficerSearchResults: get(['housingofficersearch', 'searchResults']) || [],
    error: get(['tasks', 'error']),
    userId: get(['user', 'profile', 'id']),
    loggedInUserId: get(['user', 'profile', 'id']),
    userFullName: get(['user', 'profile', 'fullname']),
    notification,
    heading,
    redirectToDetailsPage,
    errorMessage: get(['tasks', 'isFormError']) ? formError : serverError,
  };
};

export const mergeProps = (
  {
    arrearsId,
    breadcrumb,
    heading,
    housingOfficerSearchResults,
    notification,
    redirectToDetailsPage,
    userId,
    ...stateProps
  },
  { dispatch },
  ownProps
) => ({
  ...stateProps,
  arrearsId,
  housingOfficerSearchResults,
  clearUserSearch: () => dispatch(clearSearch()),
  onBack: () => redirectToDetailsPage(),
  onDisplayFormError: () => dispatch(taskError(true)),
  onSubmit: ({ title, description, date, ownerId }) => {
    dispatch(AddTaskItem(arrearsId, title, description, date, ownerId, userId))
      .then(() => {
        dispatch(
          addNotification({
            dataBddPrefix: 'addTask',
            lines: [{ text: `${title} ${formatting.formatDate(date)}` }],
            icon: 'success',
            notificationType: 'confirmation',
            title: notification.title,
          })
        );
        redirectToDetailsPage();
      })
      .catch(() => {});
  },
  onSearch: term => {
    if (term && term.length >= 3) {
      dispatch(search(term));
    } else if (term.length === 0 && !!housingOfficerSearchResults.length) {
      dispatch(clearSearch());
    }
  },
  onUnmount: () => {
    dispatch(clearTaskError());
    dispatch(clearSearch());
  },
  updatePageHeader: () => dispatch(updateRibbon({ title: heading, breadcrumb })),
  // TODO: remove this next line. By the looks of it is not needed anymore. However, I'm not 100% sure.
  // So I'm just leaving it here in case this later becomes a bug, in QA.
  ...ownProps,
});

export default connect(
  mapStateToProps,
  null,
  mergeProps
);
