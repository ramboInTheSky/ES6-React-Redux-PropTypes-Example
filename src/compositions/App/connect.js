import compose from 'ramda/src/compose';
import pickAll from 'ramda/src/pickAll';
import { connect } from 'react-redux';

import { displayLogin, logout } from '../../ducks/user';
import { setEntryPoint } from '../../ducks/navigation';
import { hideModal } from '../../ducks/modal';

export const mapStateToProps = pickAll(['user', 'ribbon', 'modal']);

export const mergeProps = (stateProps, { dispatch }, ownProps) => ({
  ...stateProps,
  login: () => dispatch(displayLogin()),
  logout: () => dispatch(logout()),
  modal: {
    ...stateProps.modal,
    onClose: compose(
      dispatch,
      hideModal
    ),
  },
  setEntryPoint: path => dispatch(setEntryPoint(path)),
  ...ownProps,
});

export default connect(
  mapStateToProps,
  null,
  mergeProps
);
