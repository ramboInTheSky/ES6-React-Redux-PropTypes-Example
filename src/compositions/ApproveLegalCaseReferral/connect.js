import { connect } from 'react-redux';
import path from 'ramda/src/path';

import { addNotification } from '../../ducks/notifications';
import { updateRibbon } from '../../ducks/ribbon';
import {
  cancelLegalReferral,
  getLegalReferral,
  setLegalReferralApproved,
} from '../../ducks/legalReferral';
import getHelpers from '../../util/stateHelpers';
import downloadFile from '../../util/downloadFile';

export const mapStateToProps = (state, ownProps) => {
  const { get, getBreadcrumbs } = getHelpers(state);
  const {
    arrearsDetails: { heading: arrearsDetailsHeading },
    approvelegalCaseReferral: { errorText, heading, labels, notification, subtitle },
  } = state.dictionary;

  const arrearsId = path(['match', 'params', 'arrearsId'], ownProps);
  const submissionId = path(['match', 'params', 'submissionId'], ownProps);
  const redirectToDetailsPage = () => ownProps.history.push(`/arrears-details/${arrearsId}`);

  return {
    activePause: get(['pause', 'activePause']),
    arrearsId,
    breadcrumb: getBreadcrumbs(
      heading,
      {},
      {
        label: arrearsDetailsHeading,
        to: `/arrears-details/${arrearsId}`,
      }
    ),
    errorText,
    formError: get(['pause', 'error']),
    heading,
    isManager: get(['user', 'userType']) === 'manager',
    labels,
    loading: get(['legalReferral', 'loading']),
    notification,
    referralPack: get(['legalReferral', 'detail', 'referralPack']),
    submissionId,
    subtitle,
    redirectToDetailsPage,
  };
};

export const mergeProps = (
  {
    breadcrumb,
    heading,
    notification: { approvedTitle, cancelledTitle },
    submissionId,
    redirectToDetailsPage,
    ...stateProps
  },
  { dispatch },
  ownProps
) => ({
  ...stateProps,
  getLegalReferral: () => dispatch(getLegalReferral(submissionId)),
  heading,
  onBack: () => redirectToDetailsPage(),
  onSubmit: noteText =>
    dispatch(
      noteText
        ? cancelLegalReferral(submissionId, { noteText })
        : setLegalReferralApproved(submissionId)
    )
      .then(() => {
        redirectToDetailsPage();
        dispatch(
          addNotification({
            dataBddPrefix: 'approveLegalReferral',
            icon: 'success',
            notificationType: 'confirmation',
            title: noteText ? cancelledTitle : approvedTitle,
          })
        );
      })
      .catch(() => {}),
  updatePageHeader: () => dispatch(updateRibbon({ title: heading, breadcrumb })),
  downloadFile: (uri, filename) => downloadFile(uri, filename),
  ...ownProps,
});

export default connect(
  mapStateToProps,
  null,
  mergeProps
);
