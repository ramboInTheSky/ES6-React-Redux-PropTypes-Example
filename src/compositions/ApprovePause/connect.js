import { connect } from 'react-redux';
import path from 'ramda/src/path';
import { formatting } from 'nhh-styles';
import format from 'string-format';

import { addNotification } from '../../ducks/notifications';
import { updateRibbon } from '../../ducks/ribbon';
import { approvePause, getPause, getPauseRules } from '../../ducks/pause';
import getHelpers from '../../util/stateHelpers';

export const mapStateToProps = (state, ownProps) => {
  const { get, getBreadcrumbs } = getHelpers(state);
  const {
    approvePause: { errorText, labels, notification, pauseRequest, pauseRequestDetails },
    arrearsDetails: { heading: arrearsDetailsHeading },
  } = state.dictionary;

  const arrearsId = path(['match', 'params', 'arrearsId'], ownProps);
  const pauseId = path(['match', 'params', 'pauseId'], ownProps);
  const redirectToDetailsPage = () => ownProps.history.push(`/arrears-details/${arrearsId}`);

  return {
    activePause: get(['pause', 'activePause']),
    arrearsId,
    breadcrumb: getBreadcrumbs(
      pauseRequest,
      {},
      {
        label: arrearsDetailsHeading,
        to: `/arrears-details/${arrearsId}`,
      }
    ),
    errorText,
    formError: get(['pause', 'error']),
    heading: pauseRequestDetails,
    labels,
    loading: get(['pause', 'loading']),
    notification,
    pauseId,
    redirectToDetailsPage,
    rules: get(['pause', 'pauseRules']),
    userId: get(['user', 'profile', 'id']),
  };
};

export const mergeProps = (
  {
    activePause,
    arrearsId,
    breadcrumb,
    heading,
    notification: { approvedTitle, cancelledTitle, line1, line2 },
    pauseId,
    redirectToDetailsPage,
    userId,
    ...stateProps
  },
  { dispatch },
  ownProps
) => ({
  ...stateProps,
  activePause,
  getPause: () =>
    dispatch(getPause(arrearsId, pauseId)).then(() => dispatch(getPauseRules(arrearsId, userId))),
  onBack: () => redirectToDetailsPage(),
  onSubmit: reason =>
    dispatch(approvePause(arrearsId, pauseId, reason))
      .then(() => {
        redirectToDetailsPage();
        dispatch(
          addNotification({
            dataBddPrefix: 'approvePause',
            lines: reason
              ? null
              : [
                  { text: format(line1, { date: formatting.formatDate(activePause.endDate) }) },
                  { text: line2 },
                ],
            icon: 'success',
            notificationType: 'confirmation',
            title: reason ? cancelledTitle : approvedTitle,
          })
        );
      })
      .catch(() => {}),
  updatePageHeader: () => dispatch(updateRibbon({ title: heading, breadcrumb })),
  ...ownProps,
});

export default connect(
  mapStateToProps,
  null,
  mergeProps
);
