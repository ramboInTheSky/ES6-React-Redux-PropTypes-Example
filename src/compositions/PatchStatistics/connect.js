import { connect } from 'react-redux';

import { getArrearsStatistics } from '../../ducks/arrears';
import getHelpers from '../../util/stateHelpers';

export const mapStateToProps = state => {
  const { get, getString } = getHelpers(state);

  return {
    errorText: getString(['patchStatistics', 'errorText']),
    heading: getString(['patchStatistics', 'heading']),
    labels: getString(['patchStatistics', 'labels']),
    loading: get(['arrears', 'loadingArrearsStats']),
    arrearsStats: get(['arrears', 'stats']),
    patches: get(['patch', 'patchList']),
  };
};

export const mergeProps = (stateProps, { dispatch }, ownProps) => ({
  ...stateProps,
  getArrearsStatistics: patch => dispatch(getArrearsStatistics(patch)),
  ...ownProps,
});

export default connect(
  mapStateToProps,
  null,
  mergeProps
);
