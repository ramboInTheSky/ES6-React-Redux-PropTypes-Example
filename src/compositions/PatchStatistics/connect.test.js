import { mergeProps, mapStateToProps } from './connect';
import * as arrearsActions from '../../ducks/arrears';
import state from '../../../__mocks__/state';

describe('PatchStatistics connector', () => {
  let dispatchProps;
  let mockState;
  let mapStateToPropsResult;

  beforeEach(() => {
    mockState = state();
  });

  describe('mapStateToProps', () => {
    beforeEach(() => {
      mapStateToPropsResult = mapStateToProps(mockState);
    });
    it('generates the correct props', () => {
      expect(mapStateToPropsResult).toMatchSnapshot();
    });

    describe('mergeProps', () => {
      let result;
      beforeEach(() => {
        dispatchProps = {
          dispatch: jest.fn(() => ({
            then: cb => {
              cb();
              return { catch: () => {} };
            },
          })),
        };
        arrearsActions.getArrearsStatistics = jest.fn();
        result = mergeProps({ ...mapStateToPropsResult }, dispatchProps);
      });

      afterEach(() => {
        jest.clearAllMocks();
      });

      it('generates the correct props', () => {
        expect(result).toMatchSnapshot();
      });

      it('performs the correct actions for getArrearsStatistics', () => {
        const patch = 'providedPatch';
        result.getArrearsStatistics(patch);
        expect(dispatchProps.dispatch).toHaveBeenCalled();
        expect(arrearsActions.getArrearsStatistics).toHaveBeenCalledWith(patch);
      });
    });
  });
});
