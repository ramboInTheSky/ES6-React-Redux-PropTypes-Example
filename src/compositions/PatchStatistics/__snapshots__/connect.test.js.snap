// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PatchStatistics connector mapStateToProps generates the correct props 1`] = `
Object {
  "arrearsStats": Object {},
  "errorText": Object {
    "multiplePatchesNotAllowed": "Patch stats not available when multiple patches selected",
    "noPatch": "Patch stats not available when no patch added",
  },
  "heading": "My patch stats",
  "labels": Object {
    "NOSPs": "{NOSPs} NOSPs to be served",
    "accountsInArrears": "With {accounts} accounts in arrears",
    "accountsWithHBUCSuspended": "{accounts} accounts with HB/UC suspended",
    "appointmentsToBeMade": "{appointments} appointments to be made",
    "liveReferralsToWBTeam": "{referrals} live referrals to W&B team",
    "openTenancyReferrals": "{referrals} open tenancy support referrals",
    "outstandingLegalReferrals": "{legalReferrals} legal referrals outstanding",
    "percentageOfRentCharged": "{percentage}% in arrears of total rent charged",
    "totalArrears": "Total arrears",
  },
  "loading": false,
  "patches": Array [
    Object {
      "fullname": "Test_1 HO",
      "name": "Test_1 HO",
      "patchName": "PTCPRH01",
      "src": "https://url.to/profile/image.jpg",
    },
  ],
}
`;

exports[`PatchStatistics connector mapStateToProps mergeProps generates the correct props 1`] = `
Object {
  "arrearsStats": Object {},
  "errorText": Object {
    "multiplePatchesNotAllowed": "Patch stats not available when multiple patches selected",
    "noPatch": "Patch stats not available when no patch added",
  },
  "getArrearsStatistics": [Function],
  "heading": "My patch stats",
  "labels": Object {
    "NOSPs": "{NOSPs} NOSPs to be served",
    "accountsInArrears": "With {accounts} accounts in arrears",
    "accountsWithHBUCSuspended": "{accounts} accounts with HB/UC suspended",
    "appointmentsToBeMade": "{appointments} appointments to be made",
    "liveReferralsToWBTeam": "{referrals} live referrals to W&B team",
    "openTenancyReferrals": "{referrals} open tenancy support referrals",
    "outstandingLegalReferrals": "{legalReferrals} legal referrals outstanding",
    "percentageOfRentCharged": "{percentage}% in arrears of total rent charged",
    "totalArrears": "Total arrears",
  },
  "loading": false,
  "patches": Array [
    Object {
      "fullname": "Test_1 HO",
      "name": "Test_1 HO",
      "patchName": "PTCPRH01",
      "src": "https://url.to/profile/image.jpg",
    },
  ],
}
`;
